keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,171 @@
+
+### Goals
+- Support memory qos for pod qos classes
+
+### Non-Goals
+- additional qos design
+- support other resources qos
+
+## Proposal
+Use memory controller interfaces of cgroups v2 to support memory qos of pod qos class.
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 and `memory.max` in cgroups v2 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs.
+
+Above implement has following disadvantages:
+- `Guaranteed` pod memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` pod memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
+
+Currently we only use `memory.limit_in_bytes = sum(pod.spec.containers.resources.limits[memory])` to limit memory hardly in cgroups v1 and `memory.max = sum(pod.spec.containers.resources.limits[memory])` in cgroups v2. And `resources.requests[memory]` is not yet used neither cgroups v1 nor cgroups v2.
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup’s memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a “soft guarantee” that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup’s memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to retain request memory for `Guaranteed` pod. And for `Burstable` pod, it uses `memory.low` to achieve best-effort memory protection by summing container `requests.memory` in pod. We also set `memory.max` to do hard limit to make OOM killer work.
```

Comment:
I like this a lot.

Is there any text you can add about the memory manager and topology manager mappings?  Maybe Kevin has thoughts (@klueska )

Reply:
`memory.min` for guaranteed at container level and `memory.low` for burstable at the pod level makes sense :+1: 

It isn't clear to me where you are proposing to set `memory.max`.

Maybe a table  with examples of request/limits and the pod class and what we set exactly at the container and pod level will be useful to add here.

Reply:
> I like this a lot.
> 
> Is there any text you can add about the memory manager and topology manager mappings? Maybe Kevin has thoughts (@klueska )

ok, I will update.

Reply:
> `memory.min` for guaranteed at container level and `memory.low` for burstable at the pod level makes sense 👍
> 
> It isn't clear to me where you are proposing to set `memory.max`.
> 
Actually, `limits.memory` is converted to `memory.max` in cgroups v2 by container runtime which in cgroups v1 is `memory.limit_in_bytes`. This proposal does not do this conversion. What I mean here is to provide more comprehensive information.

> Maybe a table with examples of request/limits and the pod class and what we set exactly at the container and pod level will be useful to add here.

Reply:
> > I like this a lot.
> > Is there any text you can add about the memory manager and topology manager mappings? Maybe Kevin has thoughts (@klueska )
> 
> ok, I will update.

Updated.

Reply:
> ok, similar table is at design details. I will update an example table to clarify the rules.

Updated.

Reply:
> I like this a lot.
> 
> Is there any text you can add about the memory manager and topology manager mappings? Maybe Kevin has thoughts (@klueska )


> Maybe a table with examples of request/limits and the pod class and what we set exactly at the container and pod level will be useful to add here.

Updated.